dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java

```diff
@@ -124,70 +157,98 @@ public static DependencyGraph getCompleteDependencies(Artifact artifact)
-        try {
-          child = resolveCompileTimeDependencies(child.getArtifact());
-          // somehow we've got an infinite recursion here
-          // requires equals
-          if (path.contains(child)) {
-            System.err.println("Infinite recursion resolving " + current);
-            System.err.println("Likely cycle in " + forPath);              
-            System.err.println("Child " + child);
-          } else {
-            fullPreorder((Stack<DependencyNode>) path.clone(), child, graph);
+  }
+
+  /**
+   * Traverses dependency tree in level-order (breadth-first search) and stores {@link
+   * DependencyPath} instances corresponding to tree nodes to {@link DependencyGraph}. When
+   * resolveFullDependency flag is true, then it resolves the dependency of the artifact of the each
+   * node in the dependency tree; otherwise it just follows the given dependency tree starting with
+   * firstNode.
+   *
+   * @param firstNode node to start traversal
+   * @param graph graph to store {@link DependencyPath} instances
+   * @param resolveFullDependency flag to resolve dependency for each node in the tree. Useful for
+   *     building a complete tree of dependencies including <i>provided</i> scope
+   * @throws DependencyCollectionException when there is a problem in collecting dependency. This
+   *     happens only when resolveFullDependency is true
+   * @throws DependencyResolutionException when there is a problem in resolving dependency. This
+   *     happens only when resolveFullDependency is true
+   */
+  private static void levelOrder(
+      DependencyNode firstNode, DependencyGraph graph, boolean resolveFullDependency)
+      throws DependencyCollectionException, DependencyResolutionException {
+    Queue<LevelOrderQueueItem> queue = new ArrayDeque<>();
+    queue.add(new LevelOrderQueueItem(firstNode, new Stack<>()));
+    while (!queue.isEmpty()) {
+      LevelOrderQueueItem item = queue.poll();
+      DependencyNode dependencyNode = item.dependencyNode;
+      DependencyPath forPath = new DependencyPath();
+      Stack<DependencyNode> parentNodes = item.parentNodes;
+      parentNodes.forEach(parentNode -> forPath.add(parentNode.getArtifact()));
+      if (dependencyNode.getArtifact() != null) {
+        // When requesting dependencies of 2 or more artifacts, root DependencyNode's artifact is
+        // set to null
+        forPath.add(dependencyNode.getArtifact());
+        if (resolveFullDependency && parentNodes.contains(dependencyNode)) {
+          // TODO: change to logger
+          System.err.println("Infinite recursion resolving " + dependencyNode);
+          System.err.println("Likely cycle in " + parentNodes);
+          System.err.println("Child " + dependencyNode);
+          return;
+        }
+        parentNodes.push(dependencyNode);
+        graph.addPath(forPath);
+
+        if (resolveFullDependency && !"system".equals(dependencyNode.getDependency().getScope())) {
```

Comment:
The system exclusion was needed for earlier, existing code. But it might be a problem for static link checking. Please file an issue to look into this.

Reply:
Created 
https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/167